tensorflow_addons/layers/crf.py

```diff
@@ -0,0 +1,485 @@
+            "bias_constraint":
+            tf.keras.constraints.serialize(self.bias_constraint)
+        }
+        base_config = super(CRF, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+    def compute_output_shape(self, input_shape):
+        output_shape = input_shape[:2]
+        return output_shape
+
+    def compute_mask(self, input_, mask=None):
+        # """
+        # Set output mask to be 1D tensor, so loss method of this class can work without error.
+        # But there is big short come:
+        # layer, loss and metrics after this layer
+        # can not access meaningful mask. Which mean they can not work correctly.
+        # User only can get correct loss and metrics value from methods of this layer.
+        # """
+        # if mask is not None:
+        #     # transform mask from shape (?, ?) to (?, )
+        #     new_mask = tf.keras.backend.any(mask, axis=1)
+        #     return new_mask
+
+        return mask
+
+    def get_negative_log_likelihood(self, y_true):
+        # TODO: remove typing cast
+        self.potentials = tf.keras.backend.cast(self.potentials, tf.float32)
+        y_true = tf.keras.backend.cast(y_true, tf.int32)
+        self.sequence_length = tf.keras.backend.cast(self.sequence_length,
+                                                     tf.int32)
+        # self.chain_kernel = tf.keras.backend.cast(self.chain_kernel,
+        #                                           tf.float32)
+
+        log_likelihood, _ = crf_log_likelihood(
+            self.potentials, y_true, self.sequence_length, self.chain_kernel)
+
+        return -log_likelihood
+
+    def get_loss(self, y_true, y_pred):
+        # we don't use y_pred, but caller pass it anyway, ignore it
+        return self.get_negative_log_likelihood(y_true)
+
+    def get_accuracy(self, y_true, y_pred):
+        judge = tf.keras.backend.cast(
+            tf.keras.backend.equal(y_pred, y_true), tf.keras.backend.floatx())
+        if self.mask is None:
+            return tf.keras.backend.mean(judge)
+        else:
+            mask = tf.keras.backend.cast(self.mask, tf.keras.backend.floatx())
+            return (tf.keras.backend.sum(judge * mask) /
+                    tf.keras.backend.sum(mask))
+
+    def _dense_layer(self, input_):
+        if self.use_kernel:
+            output = self.activation(
+                tf.keras.backend.dot(input_, self.kernel) + self.bias)
+        else:
+            output = input_
+
+        return tf.keras.backend.cast(output, self.chain_kernel.dtype)
+
+    def __call__(self, inputs, *args, **kwargs):
+        outputs = super(CRF, self).__call__(inputs, *args, **kwargs)
+
+        # A hack that add _keras_history to EagerTensor, make it more like normal Tensor
+        for tensor in tf.nest.flatten(outputs):
+            if not hasattr(tensor, '_keras_history'):
+                tensor._keras_history = (self, 0, 0)
```

Comment:
It's probably not a top priority right now, but do you have ideas/plans for the future on a better way to do it? It works alright now in most cases, but I get an error that _keras_history not found in some cases when input is unknown at the compile time and build is delayed to fit time.

Reply:
Can you tell me the case that you get the error, so I can try to resolve it? The reason I do a little hack here because, in eager execution mode, there is (as far as I know)  no way to find the layer who produced the output tensor. But CRF loss function depends on use tensor info to find the CRF layer instance. I have to add _keras_history to the tensor to make the eager tensor behave like the normal tensor generated under graph mode. The better way to do this maybe need TensorFlow core to add more information in eager tensor, or provide another way to find the layer who produced the tensor. I will try to submit an issue or PR for this.

Reply:
One way to reproduce it is subclassing:
    class CRFModel(tf.keras.Model):
            def __init__(self):
                super(CRFModel, self).__init__()
                self.layer =  CRF(5)
            def call(self, inputs):
                return self.layer(inputs)

But I believe it's also possible to get the same error with custom first layer in sequential (if there is no input layer)

Reply:
Currently, using the CRF layer in a subclassing model need a little hack or patch, I already submit a commit contains a test case to show how to hack. But the graceful way to resolve this need writes a patch to TensorFlow core.